File: src/providers/pwd.ts
import path from 'node:path';
import os from 'node:os';
import type { DataProvider, SessionData } from '../types.js';

export interface PwdData {
  name: string;
  path: string;   // full prefix path (without basename), e.g. '/Users/test/'
  smart: string;  // smart-truncated prefix (without basename), e.g. '~/t/'
}

const MAX_ABBREVIATED = 2;

function smartPrefix(cwd: string): string {
  if (cwd === '/') return '';

  const home = os.homedir();
  let p = cwd;

  // Replace home dir with ~
  if (p.startsWith(home)) {
    p = '~' + p.slice(home.length);
  }

  const parts = p.split('/');
  // 2 segments (e.g. /tmp) — just the leading slash
  if (parts.length <= 2) return parts[0] + '/';

  const first = parts[0]; // '' for absolute, '~' for home-relative
  const middle = parts.slice(1, -1);

  // 3 segments (e.g. ~/Projects/ccnow) — keep prefix as-is
  if (parts.length <= 3) return first + '/' + middle.join('/') + '/';

  // Abbreviate up to MAX_ABBREVIATED middle segments, then ellipsis if more remain
  if (middle.length <= MAX_ABBREVIATED) {
    const abbreviated = middle.map((part) => part[0] ?? '');
    return [first, ...abbreviated, ''].join('/');
  }

  const abbreviated = middle.slice(0, MAX_ABBREVIATED).map((part) => part[0] ?? '');
  return [first, ...abbreviated, '\u2026', ''].join('/');
}

export const pwdProvider: DataProvider = {
  name: 'pwd',
  async resolve(session: SessionData): Promise<PwdData> {
    const cwd = session.cwd;
    const name = cwd === '/' ? '/' : path.basename(cwd);
    const fullPrefix = cwd === '/' ? '' : path.dirname(cwd) + '/';

    return {
      name,
      path: fullPrefix,
      smart: smartPrefix(cwd),
    };
  },
};

File: internal/provider/pwd.go
package provider

import (
	"os"
	"path/filepath"
	"strings"

	"github.com/jheddings/ccglow/internal/types"
)

type pwdProvider struct{}

func (p *pwdProvider) Name() string { return "pwd" }

func (p *pwdProvider) Resolve(session *types.SessionData) (*types.ProviderResult, error) {
	cwd := session.CWD
	name := filepath.Base(cwd)
	dir := filepath.Dir(cwd)
	if dir != "/" {
		dir += "/"
	}

	return &types.ProviderResult{
		Values: map[string]any{
			"pwd": map[string]any{
				"name":  name,
				"path":  dir,
				"smart": smartPrefix(cwd),
			},
		},
	}, nil
}

func smartPrefix(cwd string) string {
	if cwd == "/" {
		return ""
	}

	home, _ := os.UserHomeDir()

	display := cwd
	if home != "" && strings.HasPrefix(cwd, home) {
		display = "~" + cwd[len(home):]
	}

	dir := filepath.Dir(display)
	if dir == "." || dir == "/" {
		return ""
	}

	// Separate the root prefix from the relative path segments
	root := ""
	rel := dir
	if strings.HasPrefix(dir, "~/") {
		root = "~/"
		rel = dir[2:]
	} else if dir == "~" {
		return "~/"
	} else if strings.HasPrefix(dir, "/") {
		root = "/"
		rel = dir[1:]
	}

	if rel == "" {
		return root
	}

	segments := strings.Split(rel, "/")

	if len(segments) <= 2 {
		return root + strings.Join(segments, "/") + "/"
	}

	// Abbreviate: first char of leading parts, then ellipsis
	var abbrev []string
	for i := 0; i < len(segments)-1 && i < 2; i++ {
		if len(segments[i]) > 0 {
			abbrev = append(abbrev, string(segments[i][0]))
		}
	}
	abbrev = append(abbrev, "\u2026")

	return root + strings.Join(abbrev, "/") + "/"
}
